File: dvbViewerController/src/main/java/org/dvbviewer/controller/data/media/MediaViewModel.java
package org.dvbviewer.controller.data.media;

import android.app.Application;
import android.arch.lifecycle.MutableLiveData;
import android.os.AsyncTask;
import android.util.Log;

import org.dvbviewer.controller.data.ApiResponse;
import org.dvbviewer.controller.data.DmsViewModel;

import java.io.IOException;
import java.util.List;

/**
 * Created by rbaun on 02.04.18.
 */

public class MediaViewModel extends DmsViewModel {

    private static final String TAG = MediaViewModel.class.getName();

    private final MediaRepository mRepository;
    private MutableLiveData<ApiResponse<List<MediaFile>>> data;

    MediaViewModel(Application application, MediaRepository repository) {
        super(application);
        mRepository = repository;
    }

    public MutableLiveData<ApiResponse<List<MediaFile>>> getMedias(final long dirid) {
        if(data == null) {
            data = new MutableLiveData<>();
            fetchMedias(dirid);
        }
        return data;
    }

    public void fetchMedias(final long dirid) {
        new AsyncTask<Void,Void,ApiResponse<List<MediaFile>>>() {

            @Override
            protected ApiResponse<List<MediaFile>> doInBackground(Void... voids) {
                try {
                    return ApiResponse.success(mRepository.getMedias(dirid));
                } catch (IOException e) {
                    Log.e(TAG, "Error loading Data", e);
                    final String message = getErrorMessage(e);
                    return ApiResponse.error(message, null);
                }
            }

            @Override
            protected void onPostExecute(ApiResponse<List<MediaFile>> response) {
                data.setValue(response);
            }
        }.execute();
    }

}

File: dvbViewerController/src/main/java/org/dvbviewer/controller/data/media/MediaViewModel.kt
package org.dvbviewer.controller.data.media

import android.app.Application
import android.util.Log
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import org.dvbviewer.controller.data.DmsViewModel
import org.dvbviewer.controller.data.api.ApiResponse

/**
 * Created by rbaun on 02.04.18.
 */

class MediaViewModel internal constructor(application: Application, private val mRepository: MediaRepository) : DmsViewModel(application) {

    private var data: MutableLiveData<ApiResponse<List<MediaFile>>>? = null

    fun getMedias(dirid: Long): MutableLiveData<ApiResponse<List<MediaFile>>> {
        if (data == null) {
            data = MutableLiveData()
            fetchMedias(dirid)
        }
        return data as MutableLiveData<ApiResponse<List<MediaFile>>>
    }

    fun fetchMedias(dirid: Long) {
        if (data == null) {
            return
        }
        viewModelScope.launch(Dispatchers.Main, CoroutineStart.DEFAULT) {
            var apiResponse: ApiResponse<List<MediaFile>> = ApiResponse.error(null, null)
            async(Dispatchers.Default) {
                apiResponse = try {
                    val mediaList = mRepository.getMedias(dirid)
                    ApiResponse.success(mediaList)
                } catch (e: Exception) {
                    Log.e(TAG, "Error getting mediafiles", e)
                    ApiResponse.error(e, null)
                }
            }.await()
            data?.value = apiResponse
        }
    }

    companion object {

        private val TAG = MediaViewModel::class.java.name
    }

}